### PR TITLE
Speedup `CumSum` for large arrays

### PR DESCRIPTION
**Pull request by neNasko1**
### Description
This PR refactors the `CPU` kernel for the `CumSum` operator. The new implementation strives to have as little indirection as possible. 

### Motivation and Context
Currently the `CumSum` operator perform very poorly in the case of 1D tensors(it was slower than a python loop). This is caused by the extensive use of the `SliceIterator`-s. 

Here is a relevant snippet:
```python
import time
import ndonnx as ndx
import onnxruntime as ort
import numpy as np
import onnx

def test_cumsum(sz):
    a = ndx.array(shape=(sz,), dtype=ndx.int64)
    b = ndx.cumsum(a)
    model = ndx.build({'a': a}, {'b': b})
    onnx.save(model, "model.onnx")

    input = np.ones(sz, np.int64)
    start = time.time()
    result = ort.InferenceSession(model.SerializeToString()).run(None, {'a': input})
    end = time.time()
    return end - start

def test_cumsum_by_hand(sz):
    input = np.ones(sz, np.int64)
    start = time.time()
    answer = [0]
    for i in input:
        answer.append(answer[-1] + i)
    end = time.time()
    return end - start

print(test_cumsum(int(1e7))) 
print(test_cumsum_by_hand(int(1e7))) 
```

Before
```console
0.9794480800628662
0.4518160820007324
```

After
```console
0.02483987808227539
0.5496008396148682
```

The `model.onnx`: 
image

The flame graph:
profile-3

### CODE REVIEW

File: onnxruntime/core/providers/cpu/math/cumsum.cc [hunk 1]
```diff
@@ -200,62 +162,82 @@ Status CumSum<T>::Compute(OpKernelContext* ctx) const {
...
-  auto slice_dims(slice_shape.GetDims());  // dim array for the slice
+  // we solve the problem by using the identity that(in the case of exclusive)
+  // 1) out[upper_dims...][0][lower_dims...] = 0
+  // 2) out[upper_dims...][i][lower_dims...] =
+  //      in[upper_dims...][i-1][lower_dims...] + out[upper_dims...][i-1][lower_dims...]
+  // we loop through the [upper_dims...] and start applying the identity in each slice
+  // since the [lower_dims...] are adjecent in memory, so we can add them like vectors
 
-  std::vector<int64_t> steps(onnxruntime::narrow<size_t>(rank), 1);  // steps for the slice -- always set to 1
+  const auto dim = input->Shape()[axis];  // dimension size for the axis
```
Comment (snnn) [inline onnxruntime/core/providers/cpu/math/cumsum.cc:172]: You need to keep the `onnxruntime::narrow<size_t>(...)` cast.

**Comment by neNasko1** [comment]
@snnn can we rerun the pipelines again? Seems like there was some conversion issue that existed.

**Comment by neNasko1** [comment]
@snnn all cpu tests are passing, I don't think that the CI failures are caused by this change as they are only on the CUDA and DML pipelines and there is this line in the logs:

```console
1: Run failed but expected success: CUDA failure 702: the launch timed out and was terminated ; GPU=0 ; hostname=7e50a0e1c000000 ; file=D:\a\_work\1\s\onnxruntime\core\providers\cuda\gpu_data_transfer.cc ; line=73 ; expr=cudaMemcpyAsync(dst_data, src_data, bytes, cudaMemcpyDeviceToHost, static_cast<cudaStream_t>(stream.GetHandle())); 
```

Can you comment if there is something left for me to do?

**Comment by neNasko1** [comment]
I think the test was failing because the tensor was too big and it timed out. Can we rerun one last time?

**Review by snnn — APPROVED** [review]

**Comment by snnn** [comment]
Thank you!

**Pull request merged**